src/analysis/reaching-definitions-transfer-function.h

```diff
@@ -0,0 +1,178 @@
+#ifndef wasm_analysis_reaching_definitions_transfer_function_h
+#define wasm_analysis_reaching_definitions_transfer_function_h
+
+#include "ir/find_all.h"
+#include "ir/local-graph.h"
+#include "visitor-transfer-function.h"
+
+namespace wasm::analysis {
+
+// A state contains all the "live" LocalSet instances at some particular
+// point in a program. The state tracks every LocalSet instance, not just
+// ones which influence a particular local index.
+
+// In addition to LocalSet expressions modifying a local index's value, we
+// must also account for the fact that the value of a local index could be
+// the default value at initialization or its value passed in as a parameter.
+// As a result, we use a fictitious LocalSet object held by the transfer
+// function to represent the local index obtaining its current value on
+// initialization or parameter passing.
+
+// Each local index gets an individual fictitious LocalSet, because the state
+// tracks all local indices, and setting one local index should not affect the
+// initial value of another.
+
+// When collecting results, the transfer function takes the states and converts
+// it into a map of LocalGets to LocalSets which affect it. The fictitious
+// inital value LocalSetes will be converted to nullptrs.
+
+class ReachingDefinitionsTransferFunction
+  : public VisitorTransferFunc<ReachingDefinitionsTransferFunction,
+                               FinitePowersetLattice<LocalSet*>,
+                               AnalysisDirection::Forward> {
+
+  // Number of locals in the function.
+  size_t numLocals;
+
+  // Maps a local index to a list of LocalSets that modify the index's value.
+  // The most common case is to have a single set; after that, to be a phi of 2
+  // items, so we use a small set of size 2 to avoid allocations there.
+  std::unordered_map<Index, SmallVector<LocalSet*, 2>> indexSetses;
+
+  // LocalGraph members we need to update.
+  LocalGraph::GetSetses& getSetses;
+
+  // Fictitious LocalSet objects to reprsent a local index obtaining its value
+  // from its default initial value or parameter value.
+  std::vector<LocalSet> fakeInitialValueSetses;
```

Comment:
The suffix "es" in "Setses" is a (silly) way to indicate "a group of sets". That is, we go from one item to a set of items to a group of sets of items. For lack of a better term, we've been using "setses" for that in LocalGraph.

Specifically, `GetSetses getSetses` there means: each get has a set of sets for it, and this is the map of each of those get+set of sets items. Or, to put it another way, "getSets" sounds like "a map of gets to sets", but we want a map of gets to sets of sets, so we need to be even more plural.

This line, however, looks like a normal vector of sets, so it could be `fakeInitialValueSets`? Or am I missing something?

Reply:
No I misunderstood `Setses` as being a more unambiguous way to distinguish a local.set from a set (i.e. group) of items. I'll change it into `fakeInitialValueSets`.